### ISSUE TITLE
Migration from QVCSPro?

### ISSUE THREAD

**Issue by nurbles** [jimv39/qvcsos#13]
Is there any way to migrate from QVCSPro (yes, we've been using it for over **20 years**) to QVCS Enterprise?

If yes, could you please direct me to the instructions, thanks!

If not... was the QVCSPro source code ever made available after development on that product was stopped?

We are interested in moving to Visual Studio 2022, but Microsoft has abandoned support for 32 bit source control products.

**Comment by jimv39** [comment]
Hmmm. I have a migration test that conceivably could do the trick. If you pull the latest from GitHub, the migration test is called ArchiveMigrationTest in the qvcse-server project. Its purpose is to populate a postgres database with file revision data from old style QVCS archive files. I put it together a while ago, and it seemed to work, but I haven't done anything with it for a while.

If you want to continue to use your old QVCS archive files, the version posted here on the master branch might work.

The QVCS/QVCS-Pro code was never made public as it contained licensed code that I didn't have rights to share.

**Comment by nurbles** [comment]
Thanks, your thoughts will give me alternates to try if my first attempt fails...  

While awaiting your response I started creating a 64bit SCC DLL that will communicate with a 32bit EXE that will access QvcsScc.dll.  If all goes as hoped, I will provide a path for VS2022 to access QVCSPro 3.10!  (smile)  If there's anything you are willing and able to share about the QVCSScc.dll implementation, it might help -- in the absence of specifics, I'm trying to create a generic thing that would support any 32bit SCC DLL using up through v1.3 of the interface.

FWIW, I think I've finished the DLL (client) and am getting started on the EXE (server) part of my package.

Regardless of any additional replies or information, THANK YOU for responding!

**Comment by jimv39** [comment]
Good luck. 

The latest QVCS Enterprise does not support Microsoft SCC IDE integration, so it sounds like your strategy offers the best chance at retaining that support.

**Comment by nurbles** [comment]
Since you left this open... (smile) I'm going to ask two more questions:
1. Can you help me to understand the LPCMDOPTS parameter that the IDE passes to QVCS?  It is actually a void pointer and Microsoft's documentation implies it is something for the SCC provider's use. Since I'm trying to act as a wedge between the 64 bit IDE and the 32 QVCS, I needs to know _what it points to_ so that I can transfer the data correctly.  Since you probably don't remember...question 2...
2. Is the source available for QVCS Pro and if so, how can I get it?

**Comment by jimv39** [comment]
You can get the QVCS Scc source here:

https://drive.proton.me/urls/551H7P57FR#ha2cy1wEQWp8

I'll leave it there a short while. Let me know when you get it.

The LPCMDOPTS are pointers to command options for the given command... e.g. a 'checkout' operation has options like whether the checkout is 'by label', or 'by revision', and if 'by revision' what the revision value is, etc. Basically, these are sort of wrappers around the various command line options for the associated QVCS command line tool... See the source for what they look like.

**Comment by nurbles** [comment]
THANKS!  I was actually going to delete my request this morning because the fact that I didn't actually _need_ to know what LPCMDOPTS pointed to, I just needed to save the value from QVCS for translation into something I could give/receive from the 64-bit IDE.

HOWEVER, the source will help me answer any other questions I may have.  I promise to stop bothering you now that I have downloaded the source.  Go ahead and close this issue as resolved to keep me from being tempted!  (smile)

Thanks again, you're great!
